File: leetcode/product_arr.py
def productExceptSelf(nums: list[int]) -> list[int]:

    length = len(nums)
    res = [1] * length
    preI = 1
    postI = 1
    i = 0
    while length > i:
        res[i] *= preI
        preI *= nums[i]
        res[length - i - 1] *= postI
        postI *= nums[length - i - 1]
        i += 1
    return res


print(productExceptSelf([1, 2, 3, 4]))

File: leetcode/product_arr.js
let productExceptSelf = function (nums) {
    let len = nums.length;
    let result = new Array(len).fill(1);
    let preI = 1;
    let postI = 1;
    let i = 0;

    while (len > i) {
        result[i] *= preI;
        preI *= nums[i];
        result[len - i - 1] *= postI;
        postI *= nums[len - i - 1];
        i++;
    }
    return result;
}
console.log(productExceptSelf([1, 2, 3, 4]));
